File: hdf-diff/test/sbom.test.ts
import { describe, it, expect } from 'vitest';
import { diffSboms } from '../src/sbom.js';

const oldCycloneDX = JSON.stringify({
  bomFormat: 'CycloneDX',
  specVersion: '1.5',
  version: 1,
  components: [
    { type: 'library', name: 'lodash', version: '4.17.20', purl: 'pkg:npm/lodash@4.17.20' },
    { type: 'library', name: 'express', version: '4.18.0', purl: 'pkg:npm/express@4.18.0' },
    { type: 'library', name: 'old-lib', version: '1.0.0', purl: 'pkg:npm/old-lib@1.0.0' },
  ],
});

const newCycloneDX = JSON.stringify({
  bomFormat: 'CycloneDX',
  specVersion: '1.5',
  version: 1,
  components: [
    { type: 'library', name: 'lodash', version: '4.17.21', purl: 'pkg:npm/lodash@4.17.21' },
    { type: 'library', name: 'express', version: '4.18.0', purl: 'pkg:npm/express@4.18.0' },
    { type: 'library', name: 'new-lib', version: '2.0.0', purl: 'pkg:npm/new-lib@2.0.0' },
  ],
});

describe('diffSboms', () => {
  describe('CycloneDX comparison', () => {
    it('should detect updated packages (version change)', () => {
      const result = diffSboms(oldCycloneDX, newCycloneDX);
      const lodash = result.packageDiffs.find(d => d.name === 'lodash');
      expect(lodash).toBeDefined();
      expect(lodash!.state).toBe('updated');
      expect(lodash!.oldVersion).toBe('4.17.20');
      expect(lodash!.newVersion).toBe('4.17.21');
    });

    it('should detect unchanged packages', () => {
      const result = diffSboms(oldCycloneDX, newCycloneDX);
      const express = result.packageDiffs.find(d => d.name === 'express');
      expect(express).toBeDefined();
      expect(express!.state).toBe('unchanged');
    });

    it('should detect removed packages', () => {
      const result = diffSboms(oldCycloneDX, newCycloneDX);
      const oldLib = result.packageDiffs.find(d => d.name === 'old-lib');
      expect(oldLib).toBeDefined();
      expect(oldLib!.state).toBe('removed');
      expect(oldLib!.oldVersion).toBe('1.0.0');
    });

    it('should detect added packages', () => {
      const result = diffSboms(oldCycloneDX, newCycloneDX);
      const newLib = result.packageDiffs.find(d => d.name === 'new-lib');
      expect(newLib).toBeDefined();
      expect(newLib!.state).toBe('added');
      expect(newLib!.newVersion).toBe('2.0.0');
    });

    it('should produce correct counts', () => {
      const result = diffSboms(oldCycloneDX, newCycloneDX);
      expect(result.added).toBe(1);
      expect(result.removed).toBe(1);
      expect(result.updated).toBe(1);
      expect(result.unchanged).toBe(1);
    });

    it('should sort diffs by name', () => {
      const result = diffSboms(oldCycloneDX, newCycloneDX);
      for (let i = 1; i < result.packageDiffs.length; i++) {
        expect(result.packageDiffs[i - 1]!.name <= result.packageDiffs[i]!.name).toBe(true);
      }
    });
  });

  describe('identical SBOMs', () => {
    it('should report all packages as unchanged', () => {
      const result = diffSboms(oldCycloneDX, oldCycloneDX);
      expect(result.added).toBe(0);
      expect(result.removed).toBe(0);
      expect(result.updated).toBe(0);
      expect(result.unchanged).toBe(3);
    });
  });

  describe('empty SBOMs', () => {
    it('should produce empty result for two empty SBOMs', () => {
      const empty = JSON.stringify({
        bomFormat: 'CycloneDX',
        specVersion: '1.5',
        version: 1,
        components: [],
      });
      const result = diffSboms(empty, empty);
      expect(result.packageDiffs).toHaveLength(0);
      expect(result.added).toBe(0);
      expect(result.removed).toBe(0);
      expect(result.updated).toBe(0);
      expect(result.unchanged).toBe(0);
    });
  });

  describe('all removed', () => {
    it('should mark all packages as removed when new SBOM is empty', () => {
      const empty = JSON.stringify({
        bomFormat: 'CycloneDX',
        specVersion: '1.5',
        version: 1,
        components: [],
      });
      const result = diffSboms(oldCycloneDX, empty);
      expect(result.removed).toBe(3);
      expect(result.added).toBe(0);
    });
  });

  describe('all added', () => {
    it('should mark all packages as added when old SBOM is empty', () => {
      const empty = JSON.stringify({
        bomFormat: 'CycloneDX',
        specVersion: '1.5',
        version: 1,
        components: [],
      });
      const result = diffSboms(empty, newCycloneDX);
      expect(result.added).toBe(3);
      expect(result.removed).toBe(0);
    });
  });

  describe('SPDX format', () => {
    const oldSpdx = JSON.stringify({
      spdxVersion: 'SPDX-2.3',
      dataLicense: 'CC0-1.0',
      SPDXID: 'SPDXRef-DOCUMENT',
      packages: [
        {
          SPDXID: 'SPDXRef-lodash',
          name: 'lodash',
          versionInfo: '4.17.20',
          externalRefs: [
            { referenceCategory: 'PACKAGE-MANAGER', referenceType: 'purl', referenceLocator: 'pkg:npm/lodash@4.17.20' },
          ],
          licenseConcluded: 'MIT',
        },
        {
          SPDXID: 'SPDXRef-express',
          name: 'express',
          versionInfo: '4.18.0',
          externalRefs: [
            { referenceCategory: 'PACKAGE-MANAGER', referenceType: 'purl', referenceLocator: 'pkg:npm/express@4.18.0' },
          ],
          licenseConcluded: 'MIT',
        },
      ],
    });

    const newSpdx = JSON.stringify({
      spdxVersion: 'SPDX-2.3',
      dataLicense: 'CC0-1.0',
      SPDXID: 'SPDXRef-DOCUMENT',
      packages: [
        {
          SPDXID: 'SPDXRef-lodash',
          name: 'lodash',
          versionInfo: '4.17.21',
          externalRefs: [
            { referenceCategory: 'PACKAGE-MANAGER', referenceType: 'purl', referenceLocator: 'pkg:npm/lodash@4.17.21' },
          ],
          licenseConcluded: 'MIT',
        },
        {
          SPDXID: 'SPDXRef-axios',
          name: 'axios',
          versionInfo: '1.6.0',
          externalRefs: [
            { referenceCategory: 'PACKAGE-MANAGER', referenceType: 'purl', referenceLocator: 'pkg:npm/axios@1.6.0' },
          ],
          licenseConcluded: 'MIT',
        },
      ],
    });

    it('should parse and compare SPDX documents', () => {
      const result = diffSboms(oldSpdx, newSpdx);
      expect(result.updated).toBe(1);   // lodash
      expect(result.removed).toBe(1);   // express
      expect(result.added).toBe(1);     // axios
      expect(result.unchanged).toBe(0);
    });

    it('should extract PURL from SPDX externalRefs', () => {
      const result = diffSboms(oldSpdx, newSpdx);
      const lodash = result.packageDiffs.find(d => d.name === 'lodash');
      expect(lodash).toBeDefined();
      expect(lodash!.purl).toContain('pkg:npm/lodash');
    });

    it('should extract licenses from SPDX', () => {
      const result = diffSboms(oldSpdx, newSpdx);
      const axios = result.packageDiffs.find(d => d.name === 'axios');
      expect(axios).toBeDefined();
      expect(axios!.licenses).toContain('MIT');
    });
  });

  describe('CycloneDX licenses', () => {
    it('should extract licenses from CycloneDX components', () => {
      const withLicenses = JSON.stringify({
        bomFormat: 'CycloneDX',
        specVersion: '1.5',
        version: 1,
        components: [
          {
            type: 'library',
            name: 'pkg-a',
            version: '1.0.0',
            licenses: [{ license: { id: 'MIT' } }],
          },
        ],
      });
      const empty = JSON.stringify({
        bomFormat: 'CycloneDX',
        specVersion: '1.5',
        version: 1,
        components: [],
      });
      const result = diffSboms(empty, withLicenses);
      const pkgA = result.packageDiffs.find(d => d.name === 'pkg-a');
      expect(pkgA).toBeDefined();
      expect(pkgA!.licenses).toContain('MIT');
    });
  });

  describe('error handling', () => {
    it('should throw for invalid JSON', () => {
      expect(() => diffSboms('not json', newCycloneDX)).toThrow();
    });

    it('should throw for unrecognized SBOM format', () => {
      const badDoc = JSON.stringify({ someField: 'value' });
      expect(() => diffSboms(badDoc, badDoc)).toThrow('Unrecognized SBOM format');
    });
  });

  describe('PURL in output', () => {
    it('should include PURL strings when available', () => {
      const result = diffSboms(oldCycloneDX, newCycloneDX);
      const hasPurl = result.packageDiffs.some(d => d.purl.includes('pkg:'));
      expect(hasPurl).toBe(true);
    });
  });
});

File: hdf-cli/pkg/diff/sbom/sbom_test.go
package sbom

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

const oldCycloneDX = `{
  "bomFormat": "CycloneDX",
  "specVersion": "1.5",
  "version": 1,
  "serialNumber": "urn:uuid:00000000-0000-0000-0000-000000000001",
  "metadata": {
    "timestamp": "2025-01-01T00:00:00Z",
    "tools": [{"name": "test"}]
  },
  "components": [
    {"type": "library", "name": "lodash", "version": "4.17.20", "purl": "pkg:npm/lodash@4.17.20"},
    {"type": "library", "name": "express", "version": "4.18.0", "purl": "pkg:npm/express@4.18.0"},
    {"type": "library", "name": "old-lib", "version": "1.0.0", "purl": "pkg:npm/old-lib@1.0.0"}
  ]
}`

const newCycloneDX = `{
  "bomFormat": "CycloneDX",
  "specVersion": "1.5",
  "version": 1,
  "serialNumber": "urn:uuid:00000000-0000-0000-0000-000000000002",
  "metadata": {
    "timestamp": "2025-01-02T00:00:00Z",
    "tools": [{"name": "test"}]
  },
  "components": [
    {"type": "library", "name": "lodash", "version": "4.17.21", "purl": "pkg:npm/lodash@4.17.21"},
    {"type": "library", "name": "express", "version": "4.18.0", "purl": "pkg:npm/express@4.18.0"},
    {"type": "library", "name": "new-lib", "version": "2.0.0", "purl": "pkg:npm/new-lib@2.0.0"}
  ]
}`

const emptyCycloneDX = `{
  "bomFormat": "CycloneDX",
  "specVersion": "1.5",
  "version": 1,
  "serialNumber": "urn:uuid:00000000-0000-0000-0000-000000000000",
  "metadata": {"timestamp": "2025-01-01T00:00:00Z", "tools": [{"name": "test"}]},
  "components": []
}`

func TestDiffSBOMs_CycloneDX(t *testing.T) {
	result, err := DiffSBOMs([]byte(oldCycloneDX), []byte(newCycloneDX))
	require.NoError(t, err)
	require.NotNil(t, result)

	// Verify counts
	assert.Equal(t, 1, result.Added, "expected 1 added package")
	assert.Equal(t, 1, result.Removed, "expected 1 removed package")
	assert.Equal(t, 1, result.Updated, "expected 1 updated package")
	assert.Equal(t, 1, result.Unchanged, "expected 1 unchanged package")

	// Verify we have 4 diffs total
	require.Len(t, result.PackageDiffs, 4)

	// Build a lookup by name for easier assertions
	byName := make(map[string]PackageDiff)
	for _, d := range result.PackageDiffs {
		byName[d.Name] = d
	}

	// express: unchanged
	express, ok := byName["express"]
	require.True(t, ok, "expected express in diffs")
	assert.Equal(t, "unchanged", express.State)

	// lodash: updated
	lodash, ok := byName["lodash"]
	require.True(t, ok, "expected lodash in diffs")
	assert.Equal(t, "updated", lodash.State)
	assert.Equal(t, "4.17.20", lodash.OldVersion)
	assert.Equal(t, "4.17.21", lodash.NewVersion)

	// old-lib: removed
	oldLib, ok := byName["old-lib"]
	require.True(t, ok, "expected old-lib in diffs")
	assert.Equal(t, "removed", oldLib.State)
	assert.Equal(t, "1.0.0", oldLib.OldVersion)

	// new-lib: added
	newLib, ok := byName["new-lib"]
	require.True(t, ok, "expected new-lib in diffs")
	assert.Equal(t, "added", newLib.State)
	assert.Equal(t, "2.0.0", newLib.NewVersion)
}

func TestDiffSBOMs_DeterministicOrder(t *testing.T) {
	result, err := DiffSBOMs([]byte(oldCycloneDX), []byte(newCycloneDX))
	require.NoError(t, err)

	// Should be sorted by name
	for i := 1; i < len(result.PackageDiffs); i++ {
		assert.True(t, result.PackageDiffs[i-1].Name <= result.PackageDiffs[i].Name,
			"diffs should be sorted by name: %s should come before %s",
			result.PackageDiffs[i-1].Name, result.PackageDiffs[i].Name)
	}
}

func TestDiffSBOMs_IdenticalSBOMs(t *testing.T) {
	result, err := DiffSBOMs([]byte(oldCycloneDX), []byte(oldCycloneDX))
	require.NoError(t, err)
	require.NotNil(t, result)

	assert.Equal(t, 0, result.Added)
	assert.Equal(t, 0, result.Removed)
	assert.Equal(t, 0, result.Updated)
	assert.Equal(t, 3, result.Unchanged)
}

func TestDiffSBOMs_EmptySBOMs(t *testing.T) {
	result, err := DiffSBOMs([]byte(emptyCycloneDX), []byte(emptyCycloneDX))
	require.NoError(t, err)
	require.NotNil(t, result)
	assert.Empty(t, result.PackageDiffs)
	assert.Equal(t, 0, result.Added)
	assert.Equal(t, 0, result.Removed)
	assert.Equal(t, 0, result.Updated)
	assert.Equal(t, 0, result.Unchanged)
}

func TestDiffSBOMs_InvalidInput(t *testing.T) {
	_, err := DiffSBOMs([]byte("not json"), []byte(newCycloneDX))
	assert.Error(t, err, "expected error for invalid old SBOM")
	assert.Contains(t, err.Error(), "failed to parse old SBOM")

	_, err = DiffSBOMs([]byte(oldCycloneDX), []byte("not json"))
	assert.Error(t, err, "expected error for invalid new SBOM")
	assert.Contains(t, err.Error(), "failed to parse new SBOM")
}

func TestDiffSBOMs_PurlInOutput(t *testing.T) {
	result, err := DiffSBOMs([]byte(oldCycloneDX), []byte(newCycloneDX))
	require.NoError(t, err)

	// At least some diffs should have non-empty PURL values
	hasPurl := false
	for _, d := range result.PackageDiffs {
		if d.Purl != "" {
			hasPurl = true
			break
		}
	}
	assert.True(t, hasPurl, "expected at least one diff with a non-empty PURL")
}

func TestDiffSBOMs_AllRemoved(t *testing.T) {
	result, err := DiffSBOMs([]byte(oldCycloneDX), []byte(emptyCycloneDX))
	require.NoError(t, err)
	assert.Equal(t, 3, result.Removed)
	assert.Equal(t, 0, result.Added)
	assert.Equal(t, 0, result.Updated)
	assert.Equal(t, 0, result.Unchanged)
}

func TestDiffSBOMs_AllAdded(t *testing.T) {
	result, err := DiffSBOMs([]byte(emptyCycloneDX), []byte(newCycloneDX))
	require.NoError(t, err)
	assert.Equal(t, 0, result.Removed)
	assert.Equal(t, 3, result.Added)
	assert.Equal(t, 0, result.Updated)
	assert.Equal(t, 0, result.Unchanged)
}
